File: UsbEnabler/UsbEnabler/FileSaver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Threading;

namespace UsbEnabler
{
    class FileSaver
    {

        public static void Init()
        {
            Logger.Instance.Write(LogModule.FileSaver, "Process init");
            try
            {
                FileSaver saver = new FileSaver();
                saver.Save();
            }
            catch (ThreadAbortException ex)
            {
                Logger.Instance.Write(LogModule.FileSaver, ex.ToString());
            }
        }

        public void Save()
        {
            Logger.Instance.Write(LogModule.FileSaver, "Saving started at " + DateTime.Now.ToString());
            Config cfg = Config.Instance();
            string storePath = cfg.StorePath + @"\" + System.Environment.MachineName + @"\";

            Logger.Instance.Write(LogModule.FileSaver, "Storage path " + storePath);
            while (true)
            {
                try
                {
                    string file = FileQueue.Files.Dequeue();
                    if (file != null)
                    {
                        string destFile = GetDestinationPath(storePath, file);
                        try
                        {
                            FileHelper.EnsurePath(destFile);

                            System.IO.File.Copy(file, destFile, true);
                            Logger.Instance.Write(LogModule.FileSaver, file + " saved to " + destFile);
                        }
                        catch (Exception ex) 
                        {
                            Logger.Instance.Write(LogModule.FileSaver, "Error saving " + file + " saved to " + destFile);
                            Logger.Instance.Write(LogModule.FileSaver, ex.ToString());
                        }
                    }
                }
                catch (InvalidOperationException ) // deque failed, no more items in queue
                {
                    if (FileQueue.ScanComplete == true)
                    {
                        Logger.Instance.Write(LogModule.FileSaver, "All item copied");
                        Logger.Instance.Write(LogModule.FileSaver, "... save ended at " + DateTime.Now.ToString());
                        break;
                    }
                    else
                        Thread.Sleep(new TimeSpan(0, 0, 1)); // wait 5 sec for more files to scan
                }

            }
        }

        private string GetDestinationPath(string storePath, string file)
        {
            return storePath + @"\" + file.Replace(':', '\\');
        }
    }
}

File: secureIt/FileSaver.java
import java.util.*;
import java.io.*;

public class FileSaver implements Runnable
{
	String storePath; 
	public void init()
	{
		Config cfg = Config.Instance();		
		storePath = cfg.storePath + cfg.hostName;
		FileHelper.ensureDirPath(storePath);
	}
	
	public void run()
	{
		Logger log = Logger.Instance();
		log.Write(2, "Saving started at " + new Date().toString());
		log.Write(2, "Storage path " + storePath);
		
		while (true)
		{
			while (FileStore.Files.hasItems())
			{
				String sourceFilePath = FileStore.Files.dequeue();
				if (sourceFilePath != null)
				{
					String destFilePath = getDestinationPath(storePath, sourceFilePath);
					try
					{
						FileHelper.ensureFilePath(destFilePath);

						FileHelper.copyFile(sourceFilePath, destFilePath);
						log.Write(2, sourceFilePath + " saved to " + destFilePath);
					}
					catch (Exception ex) 
					{
						log.Write(2, "Error saving " + sourceFilePath + " saved to " + destFilePath);
						log.Write(2, ex.toString());
					}
				}
			}
			
			// if file scanning complete? else wait for more files
			if (FileStore.scanComplete == true)
			{
				log.Write(2, "All item copied");
				log.Write(2, "... save ended at " + new Date().toString());
				break;
			}
			else
			{
				try
				{
					Thread.sleep(5 * 1000); // wait 5 sec for more files to scan
				}catch (InterruptedException ex)
				{
					// ignote this exception
				}
			}

		}
	}
	
	private String getDestinationPath(String storePath, String filePath)
	{
		String path = storePath + "\\" + filePath.replace(':', '\\');

		// ToDo:: Logic if path > 255 chars
		
		return path;
	}
}	